File: app/src/main/java/net/mulliken/pinenotenotebook/NoteView.java
package net.mulliken.pinenotenotebook;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;

import androidx.annotation.Nullable;

public class NoteView extends View {
    private static final String TAG = "NoteView";
    private static int displayHeight = 0;

    private Rect windowRect = new Rect();

    static {
        System.loadLibrary("pinenote");
    }

    private native void nativeOnAttachedToWindow();
    private native void nativeOnDetachedFromWindow();
    private native void nativeOnSizeChanged(int left, int top, int right, int bottom);
    private native void nativeOnWindowFocusChanged(boolean hasFocus);
    private native Bitmap nativeGetFullOverlayBitmap();
    private native Bitmap nativeGetOverlayBitmap();

    public NoteView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();

        // Get screen dimensions
        DisplayMetrics displayMetrics = new DisplayMetrics();
        getDisplay().getMetrics(displayMetrics);
        displayHeight = displayMetrics.heightPixels;

        nativeOnAttachedToWindow();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        nativeOnDetachedFromWindow();
    }

    @Override
    protected void onSizeChanged(int w, int h, int old_width, int old_height) {
        super.onSizeChanged(w, h, old_width, old_height);

        // Get the orientation of the device
        int[] l = new int[2];
        getLocationOnScreen(l);
        int left = l[0];
        int right = left + w;

        // Android considers the top of the screen as 0, so we need to invert the y coordinates
        int bottom = displayHeight - h - l[1];
        int top = displayHeight - l[1];

        windowRect.left = left;
        windowRect.top = top;
        windowRect.right = right;
        windowRect.bottom = bottom;

        nativeOnSizeChanged(left, top, right, bottom);
    }

    @Override
    public void onWindowFocusChanged(boolean hasWindowFocus) {
        super.onWindowFocusChanged(hasWindowFocus);

        nativeOnWindowFocusChanged(hasWindowFocus);

        if (hasWindowFocus) {
            nativeOnSizeChanged(
                    windowRect.left, windowRect.top, windowRect.right, windowRect.bottom);
        }
    }

    @Override
    protected void onDraw(android.graphics.Canvas canvas) {
        super.onDraw(canvas);

        // TODO: draw the note
    }

    public Bitmap getFullOverlayBitmap() {
        return nativeGetFullOverlayBitmap();
    }

    public Bitmap getOverlayBitmap() {
        return nativeGetOverlayBitmap();
    }
}

File: EinkDrawing/src/main/java/net/mulliken/einkdrawing/NoteView.kt
package net.mulliken.einkdrawing

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Rect
import android.util.AttributeSet

import android.util.DisplayMetrics
import android.view.View


class NoteView(context: Context?, attrs: AttributeSet?) : View(context, attrs) {
    companion object {
        init {
            System.loadLibrary("EinkDrawing")
        }
    }

    private val TAG = "NoteView"
    private var displayHeight = 0

    private val windowRect: Rect = Rect()

    private external fun nativeOnAttachedToWindow()
    private external fun nativeOnDetachedFromWindow()
    private external fun nativeOnSizeChanged(left: Int, top: Int, right: Int, bottom: Int)
    private external fun nativeOnWindowFocusChanged(hasFocus: Boolean)
    private external fun nativeGetFullOverlayBitmap(): Bitmap?
    private external fun nativeGetOverlayBitmap(): Bitmap?

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        // Get screen dimensions
        val displayMetrics = DisplayMetrics()
        getDisplay().getMetrics(displayMetrics)
        displayHeight = displayMetrics.heightPixels
        nativeOnAttachedToWindow()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        nativeOnDetachedFromWindow()
    }

    override fun onSizeChanged(w: Int, h: Int, old_width: Int, old_height: Int) {
        super.onSizeChanged(w, h, old_width, old_height)

        // Get the orientation of the device
        val l = IntArray(2)
        getLocationOnScreen(l)
        val left = l[0]
        val right = left + w

        // Android considers the top of the screen as 0, so we need to invert the y coordinates
        val bottom = displayHeight - h - l[1]
        val top = displayHeight - l[1]
        windowRect.left = left
        windowRect.top = top
        windowRect.right = right
        windowRect.bottom = bottom
        nativeOnSizeChanged(left, top, right, bottom)
    }

    override fun onWindowFocusChanged(hasWindowFocus: Boolean) {
        super.onWindowFocusChanged(hasWindowFocus)
        nativeOnWindowFocusChanged(hasWindowFocus)
        if (hasWindowFocus) {
            nativeOnSizeChanged(
                windowRect.left, windowRect.top, windowRect.right, windowRect.bottom
            )
        }
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        // TODO: draw the note
    }

    fun getFullOverlayBitmap(): Bitmap? = nativeGetFullOverlayBitmap()

    fun getOverlayBitmap(): Bitmap? = nativeGetOverlayBitmap()
}